data/scripts/globalevents/server_initialization.lua

```diff
@@ -0,0 +1,155 @@
+end
+
+-- Function to move expired bans to ban history
+local function moveExpiredBansToHistory()
+	local resultId = db.storeQuery("SELECT * FROM `account_bans` WHERE `expires_at` != 0 AND `expires_at` <= " .. os.time())
+	if resultId then
+		repeat
+			local accountId = Result.getNumber(resultId, "account_id")
+			db.asyncQuery("INSERT INTO `account_ban_history` (`account_id`, `reason`, `banned_at`, `expired_at`, `banned_by`) VALUES (" .. accountId .. ", " .. db.escapeString(Result.getString(resultId, "reason")) .. ", " .. Result.getNumber(resultId, "banned_at") .. ", " .. Result.getNumber(resultId, "expires_at") .. ", " .. Result.getNumber(resultId, "banned_by") .. ")")
+			db.asyncQuery("DELETE FROM `account_bans` WHERE `account_id` = " .. accountId)
+		until not Result.next(resultId)
+
+		Result.free(resultId)
+	end
+end
+
+-- Function to check and process house auctions
+local function processHouseAuctions()
+	local currentTime = os.time()
+	local resultId = db.storeQuery("SELECT `id`, `highest_bidder`, `last_bid`, " .. "(SELECT `balance` FROM `players` WHERE `players`.`id` = `highest_bidder`) AS `balance` " .. "FROM `houses` WHERE `owner` = 0 AND `bid_end` != 0 AND `bid_end` < " .. currentTime)
+	if resultId then
+		repeat
+			local house = House(Result.getNumber(resultId, "id"))
+			if house then
+				local highestBidder = Result.getNumber(resultId, "highest_bidder")
+				local balance = Result.getNumber(resultId, "balance")
+				local lastBid = Result.getNumber(resultId, "last_bid")
+				if balance >= lastBid then
+					db.query("UPDATE `players` SET `balance` = " .. (balance - lastBid) .. " WHERE `id` = " .. highestBidder)
+					house:setOwnerGuid(highestBidder)
+				end
+
+				db.asyncQuery("UPDATE `houses` SET `last_bid` = 0, `bid_end` = 0, `highest_bidder` = 0, `bid` = 0 " .. "WHERE `id` = " .. house:getId())
+			end
+		until not Result.next(resultId)
+
+		Result.free(resultId)
+	end
+end
+
+-- Function to store towns in the database
+local function storeTownsInDatabase()
+	db.query("TRUNCATE TABLE `towns`")
+
+	for i, town in ipairs(Game.getTowns()) do
+		local position = town:getTemplePosition()
+		db.query("INSERT INTO `towns` (`id`, `name`, `posx`, `posy`, `posz`) VALUES (" .. town:getId() .. ", " .. db.escapeString(town:getName()) .. ", " .. position.x .. ", " .. position.y .. ", " .. position.z .. ")")
+	end
+end
+
+-- Function to check for duplicate keys in a given variable's storage
+local function checkDuplicateStorageKeys(varName)
+	local keys = _G[varName]
+	local seen = {}
+	local duplicates = {}
+
+	for k, v in pairs(keys) do
+		if seen[v] then
+			table.insert(duplicates, v)
+		else
+			seen[v] = true
+		end
+	end
+
+	return next(duplicates) and duplicates or false
+end
+
+-- Function to check duplicated variable keys and log the results
+local function checkAndLogDuplicateKeys(variableNames)
+	for _, variableName in ipairs(variableNames) do
+		local duplicates = checkDuplicateStorageKeys(variableName)
```

Comment:
sempre que tiver um FOR, aloque a váriavel na memória uma vez fora do FOR e o laço sempre utilizará a mesma alocação, se a variável dor dentro, ele vai alocar uma váriavel por vez.

Reply:
> sempre que tiver um FOR, aloque a váriavel na memória uma vez fora do FOR e o laço sempre utilizará a mesma alocação, se a variável dor dentro, ele vai alocar uma váriavel por vez.

Isto é válido para cpp e outras linguagens que tem gerenciamento manual de memória, no caso do lua não é aplicável.